### ISSUE TITLE
SqlException (0x80131904) in QueuePeeker

### ISSUE THREAD

**Issue by ramonsmits** [Particular/NServiceBus.SqlServer#299]
It seems a SqlException (0x80131904) can occur in QueuePeeker when the endpoint instance is shutting down. I think either the peek should complete without aborting or that the possibly resulting exception(s) should be catched and properly processed.

This both happens with a local SQL2016 DB and SQL Azure.

# Exceptions

At shutdown
```
2016-09-19 12:40:01.4140|INFO|9|NServiceBus.Unicast.UnicastBus|Initiating shutdown.
2016-09-19 12:40:01.4140|WARN|6|NServiceBus.Transport.SQLServer.QueuePeeker|Sql peek operation failed
System.Data.SqlClient.SqlException (0x80131904): The request failed to run because the batch is aborted, this can be caused by abort signal sent from client, or another request is running in the same session, which makes the session busy.
Operation cancelled by user.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.CompleteAsyncExecuteReader()
   at System.Data.SqlClient.SqlCommand.InternalEndExecuteReader(IAsyncResult asyncResult, String endMethod)
   at System.Data.SqlClient.SqlCommand.EndExecuteReaderInternal(IAsyncResult asyncResult)
   at System.Data.SqlClient.SqlCommand.EndExecuteReaderAsync(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.SQLServer.TableBasedQueue.<TryPeek>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.SQLServer.QueuePeeker.<Peek>d__1.MoveNext()
ClientConnectionId:bff30c30-9638-4f6d-b859-b0dc6b2772a6
Error Number:3980,State:1,Class:16
```

Happening when doing purge at startup and then shutting down immediately.
```
2016-09-19 12:40:31.4476|WARN|9|NServiceBus.Transport.SQLServer.RepeatedFailuresOverTimeCircuitBreaker|The circuit breaker for SqlPeek will now be triggered
2016-09-19 12:40:31.4476|WARN|24|NServiceBus.Transport.SQLServer.RepeatedFailuresOverTimeCircuitBreaker|The circuit breaker for SqlPeek will now be triggered
2016-09-19 12:40:31.4476|FATAL|9|NServiceBus|Failed to peek PerformanceTest.Timeouts@[V6]
System.Data.SqlClient.SqlException (0x80131904): The request failed to run because the batch is aborted, this can be caused by abort signal sent from client, or another request is running in the same session, which makes the session busy.
Operation cancelled by user.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.CompleteAsyncExecuteReader()
   at System.Data.SqlClient.SqlCommand.InternalEndExecuteReader(IAsyncResult asyncResult, String endMethod)
   at System.Data.SqlClient.SqlCommand.EndExecuteReaderInternal(IAsyncResult asyncResult)
   at System.Data.SqlClient.SqlCommand.EndExecuteReaderAsync(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.SQLServer.TableBasedQueue.<TryPeek>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.SQLServer.QueuePeeker.<Peek>d__1.MoveNext()
ClientConnectionId:ae72657b-eeb7-47cd-809b-a2cc39e67854
Error Number:3980,State:1,Class:16```

**Comment by ramonsmits** [comment]
Some additional info. When using 3.0.0-beta0005 and core 6.0.0-beta0007 and running the same code 3 times I was not able to reproduce it for SQL Server and got it 2 times during shutdown with SQL Azure.

However with current 3.0.0-rc0001 and core 6.0.0-rc0002 it seems to always happen.

Also, it might be related to a change in core as it seems I'm now always getting this exception when I purge the queues and then shutdown. Atleast, that is how I interpret the log file:

```
2016-09-19 13:25:46.8193|INFO|4|NServiceBus.Transport.SQLServer.MessagePump|2,930.00 messages purged from queue PerformanceTest@[V6]
2016-09-19 13:25:46.8293|INFO|4|NServiceBus.Transport.SQLServer.MessagePump|0.00 messages purged from queue PerformanceTest.Timeouts@[V6]
2016-09-19 13:25:46.8293|INFO|11|NServiceBus.Transport.SQLServer.MessagePump|0.00 messages purged from queue PerformanceTest.TimeoutsDispatcher@[V6]
2016-09-19 13:25:46.8383|INFO|11|NServiceBus.Transport.SQLServer.MessagePump|0.00 messages purged from queue PerformanceTest.Retries@[V6]
2016-09-19 13:25:46.8634|INFO|6|NServiceBus.Unicast.UnicastBus|Initiating shutdown.
2016-09-19 13:25:46.8784|WARN|11|NServiceBus.Transport.SQLServer.QueuePeeker|Sql peek operation failed
System.Data.SqlClient.SqlException (0x80131904): The request failed to run because the batch is aborted, this can be caused by abort signal sent from client, or another request is running in the same session, which makes the session busy.
Operation cancelled by user.
```

**Comment by MarcinHoppe** [comment]
Did you have purge on startup set when you originally encountered this?

**Comment by ramonsmits** [comment]
@MarcinHoppe Not sure what your question is but the sequence it the same, just upgraded packages.

Sequence is:

- Start endpoint with purging enalbed, then stop.
- Optionally start send only endpoint for seeding, then stop after the seeding period thresshold is passed.
- Start endpoint and receive/process messages, then after a period stop.

**Comment by MarcinHoppe** [comment]
My question was about the first bullet point. I have my answer now :).

**Comment by tmasternak** [comment]
I've did some investigation on this one and here are my findings:
 * It looks that Ado .Net can [throw SqlException](http://stackoverflow.com/questions/18836869/sqlexception-when-canceling-async-query) when token passed to async operation has been cancelled. According to best practices an instance of `OperationCancelledException` holding reference to cancellation token should be thrown but that is not the case. 
 * I was able to reproduce this (mis)behavior with `ExecuteNonQueryAsync` but not with `ExecuteScalarAsync`. According to msdn this exception might happen only when [streaming operation timesout](https://msdn.microsoft.com/en-us/library/hh204872(v=vs.110).aspx) but looking at the stack trace it looks that cancellation can the same effect 
 * It's hard to be 100% sure that SqlException was caused by cancellation because it would require string searching that could potentially change with framework language packs.

Having all mentioned above in mind I propose to log (in debug) SqlExceptions that are thrown in when cancellation has been requested by the user. See #300 for proposal.

**Comment by MarcinHoppe** [comment]
@ramonsmits Can you check if this resolves the problem you are observing before we merge this?

**Comment by ramonsmits** [comment]
I'm just wondering, instead of catching the exception isn't the cause that something is cancelled that shouldn't be cancelled? Cancellation doesn't really reflect a graceful shutdown where you complete all current activity but don't start any new tasks.

As this is a PR I should be able to verify this with the nuget package from the build server.

**Comment by tmasternak** [comment]
Created issue for shutdown discussion https://github.com/Particular/PlatformDevelopment/issues/956

**Comment by MarcinHoppe** [comment]
Resolved by #300
